### PR TITLE
Add a UI page for masters

### LINKED ISSUE #16411: Need a master page show masters informations like worker page
**Is your feature request related to a problem? Please describe.**

Now, we have no master page, user cannot get masters state by webui, when a master lost, we cannot know the lost time.

**Describe the solution you'd like**

We can get masters info from the following objects within DefaultMetaMaster.java
```java
  private final IndexedSet<MasterInfo> mMasters =
      new IndexedSet<>(ID_INDEX, ADDRESS_INDEX);
  private final IndexedSet<MasterInfo> mLostMasters =
      new IndexedSet<>(ID_INDEX, ADDRESS_INDEX);
```

We can display the MasterInfo like the following part.

```java
public final class MasterInfo {
  /** Master's address. */
  private final Address mAddress;
  /** The id of the master. */
  private final long mId;
  /** Master's last updated time in ms. */
  private long mLastUpdatedTimeMs;
```

**Describe alternatives you've considered**

Add a master page like worker page

wecom-temp-20279f43a77ad2bfcb027b359898a751

**Urgency**

Normal

**Additional context**

No

### PR DESCRIPTION
**Pull request by lucaspeng12138**
### What changes are proposed in this pull request?

Add Master web ui page to show masters informations
> Fix #16411 

### Why are the changes needed?

1、DevOps needs to see masters cluster information
2、It can show workers and masters information directly on screen

image

### CODE REVIEW

File: core/common/src/main/java/alluxio/wire/MasterInfo.java [hunk 1]
```diff
@@ -63,16 +80,61 @@ public long getLastUpdatedTimeMs() {
...
+  public MasterInfo setLastUpdatedTimeMs(long lastUpdatedTimeMs) {
+    mLastUpdatedTimeMs = lastUpdatedTimeMs;
+    return this;
+  }
+
   @Override
   public String toString() {
     return MoreObjects.toStringHelper(this).add("id", mId).add("address", mAddress)
-        .add("lastUpdatedTimeMs", mLastUpdatedTimeMs).toString();
+            .add("lastUpdatedTimeMs", mLastUpdatedTimeMs).toString();
```
Comment (maobaolong) [inline core/common/src/main/java/alluxio/wire/MasterInfo.java:113]: Remove these unrelated changes.

File: core/common/src/main/java/alluxio/wire/MasterInfo.java [hunk 2]
```diff
@@ -9,11 +9,10 @@
  * See the NOTICE file distributed with this work for information regarding copyright ownership.
  */
 
-package alluxio.master.meta;
-
-import alluxio.wire.Address;
+package alluxio.wire;
```
Comment (maobaolong) [inline core/common/src/main/java/alluxio/wire/MasterInfo.java:12]: Please explain the reason why you move the package?

File: core/common/src/main/java/alluxio/wire/MasterInfo.java [hunk 3]
```diff
@@ -63,16 +80,61 @@ public long getLastUpdatedTimeMs() {
...
   @Override
   public String toString() {
     return MoreObjects.toStringHelper(this).add("id", mId).add("address", mAddress)
-        .add("lastUpdatedTimeMs", mLastUpdatedTimeMs).toString();
+            .add("lastUpdatedTimeMs", mLastUpdatedTimeMs).toString();
   }
 
   /**
-   * Updates the last updated time of the master in ms.
+   * Updates the last updated time of the master (in milliseconds).
```
Comment (maobaolong) [inline core/common/src/main/java/alluxio/wire/MasterInfo.java:117]: ms = milliseconds, so do you think this change is not necessary?

File: core/common/src/main/java/alluxio/wire/MasterInfo.java (hunk 2 above)
Comment (maobaolong) [reply core/common/src/main/java/alluxio/wire/MasterInfo.java:12]: Maybe we need create another MasterInfo in wire package?

File: core/common/src/main/java/alluxio/wire/MasterInfo.java [hunk 4]
```diff
@@ -24,12 +23,17 @@
 @NotThreadSafe
 public final class MasterInfo {
   /** Master's address. */
-  private final Address mAddress;
+  private Address mAddress;
```
[the remark this answers is not in the document]
Comment (lucaspeng12138) [reply core/common/src/main/java/alluxio/wire/MasterInfo.java:26]: If add setter still need to remove final.

File: core/common/src/main/java/alluxio/wire/MasterInfo.java (hunk 2 above)
Comment (lucaspeng12138) [reply core/common/src/main/java/alluxio/wire/MasterInfo.java:12]: I will change that.

File: core/common/src/main/java/alluxio/wire/MasterInfo.java (hunk 3 above)
Comment (lucaspeng12138) [reply core/common/src/main/java/alluxio/wire/MasterInfo.java:117]: Yes, but there is another test failed by accident, I use this commit to re-run it.

File: core/common/src/main/java/alluxio/wire/MasterInfo.java (hunk 1 above)
Comment (lucaspeng12138) [reply core/common/src/main/java/alluxio/wire/MasterInfo.java:113]: I will fix that

File: core/common/src/main/java/alluxio/util/webui/UIMasterInfo.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+   * @param lastUpdatedTimeMs The last heart beat in ms
+   */
+  public UIMasterInfo(String masterAddress, long id, long lastUpdatedTimeMs) {
+    mMasterAddress = masterAddress;
+    mId = id;
+    mLastUpdatedTimeMs = lastUpdatedTimeMs;
+  }
+
+  /**
+   * Get master address.
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/util/webui/UIMasterInfo.java:38]: nit: https://docs.alluxio.io/os/user/stable/en/contributor/Code-Conventions.html#javadoc-style
Conventionally we use 3rd person form instead of 2nd person

File: core/common/src/main/java/alluxio/util/webui/WebUtils.java [hunk 6]
```diff
@@ -55,5 +57,21 @@ public static NodeInfo[] generateOrderedNodeInfos(Collection<WorkerInfo> workerI
...
+   * @return the list of {@link MasterInfo} objects
+   */
+  public static MasterInfo[] generateMasterInfos(Collection<MasterInfo> masterInfos) {
+    MasterInfo[] ret = new MasterInfo[masterInfos.size()];
+    int index = 0;
+    for (MasterInfo masterInfo : masterInfos) {
+      ret[index++] = new MasterInfo().clone(masterInfo);
+    }
+
+    return ret;
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/util/webui/WebUtils.java:73]: Do you need a copy here? Just `masterInfos.toArray(new MasterInfo[0])`?

File: core/common/src/main/java/alluxio/wire/MasterInfo.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+   *
+   * @param masterInfo the masterInfo need to be cloned
+   * @return the the master information
+   */
+  public MasterInfo clone(MasterInfo masterInfo) {
+    mAddress = masterInfo.getAddress();
+    mId = masterInfo.getId();
+    mLastUpdatedTimeMs = masterInfo.getLastUpdatedTimeMs();
+    return this;
+  }
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/wire/MasterInfo.java:73]: In the 2nd or 3rd constructors, call `this(a, b, c)` instead of rewriting assignments

File: core/common/src/main/java/alluxio/wire/MasterInfo.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+   */
+  public MasterInfo setLastUpdatedTimeMs(long lastUpdatedTimeMs) {
+    mLastUpdatedTimeMs = lastUpdatedTimeMs;
+    return this;
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this).add("id", mId).add("address", mAddress)
+        .add("lastUpdatedTimeMs", mLastUpdatedTimeMs).toString();
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/wire/MasterInfo.java:126]: convert the MS to better format with `CommonUtils.convertMsToClockTime(mLastUpdatedTimeMs)`

File: core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+import java.io.Serializable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Alluxio WebUI masters information.
+ */
+@NotThreadSafe
+public final class MasterWebUIMasters implements Serializable {
+  private static final long serialVersionUID = 1320235517686567983L;
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java:24]: duplicate?

File: core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+   * @return the leader master info
+   */
+  public MasterWebUIMasters setLeaderMasterInfo(MasterInfo LeaderMasterInfo) {
+    mLeaderMasterInfo = LeaderMasterInfo;
+    return this;
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this).add("debug", mDebug)
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java:119]: 4 more spaces from the 1st line

File: core/common/src/main/java/alluxio/util/webui/WebUtils.java (hunk 6 above)
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/util/webui/WebUtils.java:73]: Therefore I don't think you need to this utility method at all?

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 11]
```diff
@@ -487,17 +489,37 @@ public List<Address> getWorkerAddresses() {
     return mWorkerConfigStore.getLiveNodeAddresses();
   }
 
+  @Override
+  public List<MasterInfo> getMasterInfoList() {
+    List<MasterInfo> masterInfoList = new ArrayList<>(mMasters.size());
+    for (MetaMasterInfo master : mMasters) {
+      masterInfoList.add(new MasterInfo(master.getId(),
+              master.getAddress(), master.getLastUpdatedTimeMs()));
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java:497]: nit: 4 more spaces from the 1st line, and same for all other code pieces

File: core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java [hunk 12]
```diff
@@ -848,6 +855,36 @@ public Response getWebUIWorkers() {
...
+  @Path(WEBUI_MASTERS)
+  public Response getWebUIMasters() {
+    return RestUtils.call(() -> {
+      MasterWebUIMasters response = new MasterWebUIMasters();
+
+      response.setDebug(Configuration.getBoolean(PropertyKey.DEBUG));
+
+      List<MasterInfo> lostMasterInfos = mMetaMaster.getLostMasterInfoList();
+      MasterInfo[] failedMasterInfos = WebUtils.generateMasterInfos(lostMasterInfos);
+      response.setFailedMasterInfos(failedMasterInfos);
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java:873]: can the response take List instead of Array? If not, `mMetaMaster.getLostMasterInfo()` can just return an array?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Added a bunch of minor comments. The feature itself looks good. Thanks for the improvements!

File: webui/master/src/containers/pages/Masters/Masters.test.tsx [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+      class: '',
+      fetchRequest: sinon.spy(() => {}),
+    };
+  });
+
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  describe('Shallow component', () => {
```
Comment (tieujason330) [inline webui/master/src/containers/pages/Masters/Masters.test.tsx:48]: - can you create a test case for when `initData.debug` is `true`?
- can you create test case for some example master data?

**Review by tieujason330 — COMMENTED** [review]
ui changes lgtm, add a few test cases to show how master ui page looks like

**Comment by jiacheliu3** [comment]
@tieujason330 PTAL the UI part of this PR. Thanks!

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 14]
```diff
@@ -487,17 +489,37 @@ public List<Address> getWorkerAddresses() {
...
+
+  @Override
+  public List<MasterInfo> getLostMasterInfoList() {
+    List<MasterInfo> masterInfoList = new ArrayList<>(mLostMasters.size());
+    for (MetaMasterInfo master : mLostMasters) {
+      masterInfoList.add(new MasterInfo(master.getId(),
+          master.getAddress(), master.getLastUpdatedTimeMs()));
+    }
+    return masterInfoList;
+  }
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java:510]: i don't really understand why the array->List conversion is needed

File: core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java [hunk 15]
```diff
@@ -22,7 +22,7 @@
  * Master information.
  */
 @NotThreadSafe
-public final class MasterInfo {
+public final class MetaMasterInfo {
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java:25]: why did we need this rename?

File: core/common/src/main/java/alluxio/util/webui/WebUtils.java [hunk 16]
```diff
@@ -22,6 +22,7 @@
  */
 @ThreadSafe
 public final class WebUtils {
+
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/util/webui/WebUtils.java:25]: remove

**Review by jiacheliu3 — COMMENTED** [review]
Overall LGTM with some minor comments. I think it's good to merge once these comments are resolved. Thanks!

File: core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java (hunk 15 above)
Comment (lucaspeng12138) [reply core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java:25]: Package Wire need a MasterInfo class, and I want to make a difference with it. Also package meta has a lot of MetaXXX class, so I change this class name.

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java (hunk 14 above)
Comment (lucaspeng12138) [reply core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java:510]: According to MasterInfo class is changed into MetaMasterInfo, there maybe some differences on params between Wire.MasterInfo and Meta.MetaMasterInfo. So this function is needed.

File: core/common/src/main/java/alluxio/util/webui/WebUtils.java (hunk 16 above)
Comment (lucaspeng12138) [reply core/common/src/main/java/alluxio/util/webui/WebUtils.java]: Fixed

**Comment by maobaolong** [comment]
@lucaspeng12138 Why there always show only one standby master? And the hostname always show localhost?

File: webui/master/src/containers/pages/Masters/Masters.test.tsx (hunk 13 above)
Comment (tieujason330) [reply webui/master/src/containers/pages/Masters/Masters.test.tsx:48]: @lucaspeng12138

File: core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java (hunk 15 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java:25]: Actually in the package `meta` all those are `MetaMaster` which correspond to the `MetaMaster` class so `MetaMaster` is one word. This `MasterInfo` actually means the metadata of masters so `Meta-Master` actually are two separate words. Therefore I prefer this stays `MasterInfo`. Let's keep the old name. Thanks!

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java (hunk 14 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java:510]: I don't really think there's change in params. Can we change this to returning a copied `MasterInfo[]`?

**Review by jiacheliu3 — COMMENTED** [review]
Added some quick comments. Rest of the code should be good. Thanks for the contribution!

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java (hunk 14 above)
Comment (lucaspeng12138) [reply core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java:510]: Actually we may add package information for master, which metaMasterInfo may not use it.

File: core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java (hunk 15 above)
Comment (lucaspeng12138) [reply core/server/master/src/main/java/alluxio/master/meta/MetaMasterInfo.java:25]: OK, I will fix that.

File: core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java [hunk 17]
```diff
@@ -848,6 +855,37 @@ public Response getWebUIWorkers() {
...
+    return RestUtils.call(() -> {
+      MasterWebUIMasters response = new MasterWebUIMasters();
+
+      response.setDebug(Configuration.getBoolean(PropertyKey.DEBUG));
+
+      MasterInfo[] failedMasterInfos = mMetaMaster.getLostMasterInfoList()
+          .toArray(new MasterInfo[0]);
+      response.setFailedMasterInfos(failedMasterInfos);
+
+      List<MasterInfo> masterInfos = mMetaMaster.getMasterInfoList();
```
Comment (kaijchen) [inline core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java:875]: Variable `masterInfos` is never used.

```suggestion
```

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 18]
```diff
@@ -487,6 +488,26 @@ public List<Address> getWorkerAddresses() {
     return mWorkerConfigStore.getLiveNodeAddresses();
   }
 
+  @Override
+  public List<alluxio.wire.MasterInfo> getMasterInfoList() {
+    List<alluxio.wire.MasterInfo> masterInfoList = new ArrayList<>(mMasters.size());
+    for (MasterInfo master : mMasters) {
+      masterInfoList.add(new alluxio.wire.MasterInfo(master.getId(),
```
Comment (kaijchen) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java:495]: It's more clear to use a copy constructor or `Object.clone()` here.
Also, is shallow copy enough for `master.mAddress`?

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 19]
```diff
@@ -487,6 +488,26 @@ public List<Address> getWorkerAddresses() {
...
+          master.getAddress(), master.getLastUpdatedTimeMs()));
```
Comment (kaijchen) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java:496]: Is shallow copy enough for `master.mAddress` here?

File: core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java (hunk 17 above)
Comment (lucaspeng12138) [reply core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java]: Fixed

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java (hunk 14 above)
Comment (lucaspeng12138) [reply core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: Fixed

**Comment by lucaspeng12138** [comment]
> @lucaspeng12138 Why there always show only one standby master? And the hostname always show localhost?

I find this is another bug in master, @kaijchen may fix it.

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java (hunk 19 above)
Comment (lucaspeng12138) [reply core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java]: yes, it is also ok here.

**Review by kaijchen — APPROVED** [review]
The latest change looks good, thanks @lucaspeng12138 for updating.

**Review by jiacheliu3 — APPROVED** [review]
LGTM except a few minor changes. Could you take another look at the code format? I'll merge once the typos are updated. Thanks a lot for the contribution!

File: core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+    return this;
+  }
+
+  /**
+   * Sets leader master info.
+   *
+   * @param LeaderMasterInfo the normal master info
+   * @return the leader master info
+   */
+  public MasterWebUIMasters setLeaderMasterInfo(MasterInfo LeaderMasterInfo) {
```
[the remark this answers is not in the document]
Comment (lucaspeng12138) [reply core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java]: Fixed

File: core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+    return this;
+  }
+
+  /**
+   * Sets failed master infos.
+   *
+   * @param FailedMasterInfos the failed master infos
+   * @return the failed master infos
+   */
+  public MasterWebUIMasters setFailedMasterInfos(MasterInfo[] FailedMasterInfos) {
```
[the remark this answers is not in the document]
Comment (lucaspeng12138) [reply core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java]: Fixed

File: core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+    return mLeaderMasterInfo;
+  }
+
+  /**
+   * Sets debug.
+   *
+   * @param Debug the debug
+   * @return the debug
+   */
+  public MasterWebUIMasters setDebug(boolean Debug) {
```
[the remark this answers is not in the document]
Comment (lucaspeng12138) [reply core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java]: Fixed

File: core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+   */
+  public MasterInfo getLeaderMasterInfo() {
+    return mLeaderMasterInfo;
+  }
+
+  /**
+   * Sets debug.
+   *
+   * @param debug the debug
+   * @return the debug
```
Comment (kaijchen) [inline core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java:77]: This javadoc is incorrect.

Comment (lucaspeng12138) [reply core/common/src/main/java/alluxio/wire/MasterWebUIMasters.java]: Fixed

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please
